### DISCUSSION TITLE
Throwing error in mutation handler is uncaught

### DISCUSSION THREAD

**Discussion by hmbrg** [TanStack/query#1784] [Q&A]
When I throw an error inside a mutation handler I still get an uncaught error.

My code looks like this: 
```javascript
  const { mutateAsync, status, isError } = useMutation(async () => {
    const { data } = await api();

    if (!data?.data?.allow_action) throw new Error('action denied');

    return data;
  });
```

I would expect the error to be caught by react-query. What am I doing wrong here?

**Comment by TkDodo** [comment]
it works for me: https://codesandbox.io/s/compassionate-frost-4zoh1?file=/pages/index.js

do you have a codesandbox that shows your issue please?

**Reply by hmbrg** [reply]
Yes it works the same way for me, but I still get this error dialog by NextJS. Is this the expected behaviour, because it doesn't happen in your example: 
![Screenshot 2021-02-11 at 14 38 36](https://user-images.githubusercontent.com/7304269/107643835-f195d000-6c76-11eb-8ecc-1615450aa242.png)

**Reply by TkDodo** [reply]
I can take a look if you create a minimal reproducible example on codesandbox

**Reply by RyanMacBern** [reply]
I'm encountering this as well. The error is not caught and appears in the console. I see this with @TkDodo 's codesandbox too.
<img width="718" alt="Screen Shot 2021-08-31 at 2 37 43 PM" src="https://user-images.githubusercontent.com/774073/131557768-971a7b3d-95c1-479d-b16e-16d8ee862ca7.png">

**Reply by TkDodo** [reply]
react-query logs errors to the console per default - it's not an uncaught error just because you see it in the console. you can also turn that off: https://react-query.tanstack.com/reference/setLogger#_top

**Comment by shayanakhlaghian** [comment]
Instead of throwing an error you can return a promise in which you call reject:

`
  if (!data?.data?.allow_action) return new Promise((_, rej)) => rej('action denied'));
`

**Comment by Maxim-Mazurok** [comment]
Having the same issue in react app, getting "Uncaught runtime errors"

Doing it the same as in docs (https://tanstack.com/query/v5/docs/framework/react/guides/mutations)

```ts
export const useSubscribeMutation = () =>
  useMutation({
    mutationFn: (email: string) => {
      return axios.post(
        new URL(`subscribe`, process.env.REACT_APP_API_BASE_URL).toString(),
        { email }
      );
    },
  });
```

**Reply by Maxim-Mazurok** [reply]
Figured it out, I was using `subscribeMutation.mutateAsync` and not doing try-catch there.

Now it works fine:
```ts
subscribeMutation
  .mutateAsync(email)
  .then(() => {
    mixpanel.track("subscribe-dialog-submitted-success", { email });
    handleClose("submitted");
  })
  .catch(() => {
    mixpanel.track("subscribe-dialog-submitted-error", { email });
  });
```

**Comment by jonahallibone** [comment]
I am also facing this problem. I guess I wrongly assumed `mutateAsync` would internally catch the error and set the error state of the hook

**Reply by TkDodo** [reply]
use `mutate`, not `mutateAsync`

**Reply by viktaur** [reply]
That depends on whether you need to chain operations afterwards

**Comment by brianjmeier** [comment]
I know this is really old but I'm having a similar issue:

I have this big mutation that composes other mutations.
I use `mutateAsync` for the internal once so error handling is cleaner I don't have to repeat the logic for each point of failure. However, this is not working. I see the error being thrown in the debugger but `onError` is never called. Thoughts @TkDodo?

```ts
export default function useCreatePrediction() {
  const router = useRouter();
  const { toast } = useToast();
  const form = usePredictionForm();
  const findOrCreateCase = useFindOrCreateCase();
  const runPrediction = useRunPrediction();

  return useMutation({
    mutationFn: async () => {
      const result = schema.safeParse(form.getValues());
      if (!result.success) throw new Error("Invalid form data");
      try {
        const newCase = await findOrCreateCase.mutateAsync(result.data);
        if (!newCase) return;
        await runPrediction.mutateAsync({
          caseId: newCase.id,
          values: result.data,
        });
        void trackOnClient().predictionCreated(form.trackingValues());
      } catch {
        throw new Error("Error creating prediction");
      }
    },
    onSuccess: () => router.push(Routes.success),
    onError: () =>
      toast({
        variant: "error",
        description:
          "There was an error creating the prediction. Please try again.",
      }),
  });
}
```

**Reply by brianjmeier** [reply]
I found the cause pretty quickly after writing this...the called will be called after all retries, what makes total sense. The issue was I had a default setting of 5 retries on all mutations, and that was being turned into 25 retries because of my nesting. Interesting side-effect of that config
